SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl

```diff
@@ -0,0 +1,598 @@
+        }
+    }
+}
+
+
+template<class DataTypes>
+void PolynomialSpringsForceField<DataTypes>::addDForce(const core::MechanicalParams* mparams, DataVecDeriv& data_df1, DataVecDeriv& data_df2,
+                                                       const DataVecDeriv& data_dx1, const DataVecDeriv& data_dx2)
+{
+    VecDeriv&        df1 = *data_df1.beginEdit();
+    VecDeriv&        df2 = *data_df2.beginEdit();
+    const VecDeriv&  dx1 =  data_dx1.getValue();
+    const VecDeriv&  dx2 =  data_dx2.getValue();
+
+    msg_info() << "[" <<  this->getName() << "]: addDforce";
+
+    Real kFactor = (Real)mparams->kFactorIncludingRayleighDamping(this->rayleighStiffness.getValue());
+    for (unsigned int index = 0; index < m_firstObjectIndices.size(); index++)
+    {
+        const JacobianMatrix& jacobMatrix = m_differential[index];
+        Deriv forceDelta = jacobMatrix * (dx2[m_secondObjectIndices[index]] - dx1[m_firstObjectIndices[index]]);
+        msg_info() << "Spring stiffness derivative: " << forceDelta;
+
+        df1[m_firstObjectIndices[index]] += forceDelta * kFactor;
+        df2[m_secondObjectIndices[index]] -= forceDelta * kFactor;
+    }
+}
+
+
+template<class DataTypes>
+void PolynomialSpringsForceField<DataTypes>::draw(const core::visual::VisualParams* vparams)
+{
+    if (!((this->mstate1 == this->mstate2)?vparams->displayFlags().getShowForceFields():vparams->displayFlags().getShowInteractionForceFields()))
+        return;
+
+    vparams->drawTool()->saveLastState();
+
+    const VecCoord& p1 =this->mstate1->read(core::ConstVecCoordId::position())->getValue();
+    const VecCoord& p2 =this->mstate2->read(core::ConstVecCoordId::position())->getValue();
+
+    const VecIndex& firstObjectIndices = d_firstObjectPoints.getValue();
+    const VecIndex& secondObjectIndices = d_secondObjectPoints.getValue();
+
+    std::vector< defaulttype::Vector3 > points;
+    for (unsigned int i = 0; i < firstObjectIndices.size(); i++)
+    {
+        const unsigned int index1 = firstObjectIndices[i];
+        points.push_back(p1[index1]);
+        const unsigned int index2 = secondObjectIndices[i];
+        points.push_back(p2[index2]);
+    }
+
+    if (d_showArrowSize.getValue() == 0 || d_drawMode.getValue() == 0)
+    {
+        vparams->drawTool()->drawLines(points, 1, d_springColor.getValue());
+    }
+    else if (d_drawMode.getValue() == 1)
+    {
+        const unsigned int numLines = points.size() / 2;
+        for (unsigned int i = 0; i < numLines; ++i) {
+            vparams->drawTool()->drawCylinder(points[2*i+1], points[2*i], d_showArrowSize.getValue(), d_springColor.getValue());
+        }
+    }
+    else if (d_drawMode.getValue() == 2)
+    {
+        const unsigned int numLines = points.size() / 2;
+        for (unsigned int i = 0; i < numLines; ++i) {
+            vparams->drawTool()->drawArrow(points[2*i+1], points[2*i], d_showArrowSize.getValue(), d_springColor.getValue());
```

Comment:
Put the drawTool() out of the loop. 
as well as all the getValue() (because this trigs the data graph update each time)